pkg/repo/chartrepo.go

```diff
@@ -249,10 +249,14 @@ func FindChartInAuthAndTLSAndPassRepoURL(repoURL, username, password, chartName,
 	if err != nil {
 		return "", err
 	}
+	// Use a random path name so it can be deleted for cleanup.
+	r.CachePath = helmpath.CachePath(name)
```

Comment:
I don't follow. `r.CachePath` is determined by `r, err := NewChartRepository(&c, getters)` above. `NewChartRepository` sets `r.CachePath` to `helmpath.CachePath("repository")`. This change overrides that cache path to instead use one that can be safely deleted after use. As far as I can tell, `r.CachePath` is inaccessible to the user. The struct is instantiated and discarded within this function.

I think this code is ugly. I think the cleanup should happen closer to where the file is actually used. Ideally, no file should be written, because the contents are immediately read back from disk and never used again.

But the nicer solution, if I'm reading the code correctly, relies on either changing a public function signature or adding a new public function (see previous commits).

Reply:
To add some more context, `FindChartInAuthAndTLSAndPassRepoURL` and `CachePath` are both public and SDK users can use them programatically.

Ignoring the `CachePath` set by the SDK user would be a functionality change. Since we know where `DownloadIndexFile` downloads to, this random location don't need to be set for cleanup.

Reply:
@mattfarina this doesn't entirely ignore the configured cachepath. It just uses a different subdirectory (a random directory name instead of `repository`). If that's an unacceptable behavior change, then yes, we'll need to use the known file name for cleanup.

> To add some more context, FindChartInAuthAndTLSAndPassRepoURL and CachePath are both public and SDK users can use them programatically.

Ignoring the CachePath set by the SDK user would be a functionality change. Since we know where DownloadIndexFile downloads to, this random location don't need to be set for cleanup.

`DownloadIndexFile` downloads two files. We know the path of the "index" file, because `DownloadIndexFile` returns that path. We can _infer_ the path of the "charts" file using `helmpath.CacheChartsFile`. But then `FindChartInAuthAndTLSAndPassRepoURL` has to assume behavior of `DownloadIndexFile`. If the implementation of `DownloadIndexFile` changes, `FindChartInAuthAndTLSAndPassRepoURL` breaks.

There are two better solutions:
1) return the chart file name so it can be cleaned up - that was rejected since it changes a public function signature
2) don't write files, and just read them into memory since they're immediately read anyway - this was rejected since it introduces a new public fuction

Both those solutions can be seen as previous commits on this branch.

Reply:
Looks like the second attempt was lost in a force-push. Basically, it introduced a `downloadIndexFileString` function so that I didn't have to write the files to disk at all. But that required creating a new public `LoadIndexString` function to parse the files.